### ISSUE TITLE
ZIO Test: TestClock.adjust can hang in certain situations

### ISSUE THREAD

**Issue by reibitto** [zio/zio#3840]
Here's a minimal example that I came up with. There's a semaphore with 0 permits, so it will block when calling `acquire` on it. After a timeout of 3 seconds it should get released though due to the cancel block.

```scala
testM("test cancel") {
  val semaphore = new Semaphore(0)
  val zio = for {
    fiber  <- effectBlockingCancelable(semaphore.acquire())(UIO(semaphore.release())).timeout(3.seconds).fork
    _      <- TestClock.adjust(3.seconds)
    result <- fiber.join
  } yield result

  assertM(zio)(isNone)
}
```

This code hangs on TestClock.adjust and the cancel block is never called.

To get a more real-life example of what I'm trying to model here, you can take a look at this test in zio-process:
https://github.com/zio/zio-process/blob/dcc8e8f1e2a4b5acb6b032b5e998236d7d38ae56/src/test/scala/zio/process/CommandSpec.scala#L86-L96

The forking code of TestClock.adjust there was a temporary workaround that worked in RC20, but it no longer works in RC21. Ideally I'd like to write something like this instead.

```scala
for {
  fiber  <- Command("sleep", "20").exitCode.timeout(5.seconds).fork
  _      <- TestClock.adjust(5.seconds)
  result <- fiber.join
} yield result
```

`Command` here is basically just doing this under the hood:

```scala
effectBlockingCancelable(process.waitFor())(UIO(process.destroy()))
```

**Comment by joroKr21** [comment]
We had a similar problem and it looks like there is a race condition between calling `ZIO.sleep` which registers a suspension in the `TestClock` and calling `TestClock.adjust` which runs it. If `TestClock.adjust` is called before the suspension is registered, then it runs nothing and we sleep forever.

Which begs the question about usability of `TestClock` - that's really a chicken and egg problem. How are we supposed to ensure that `ZIO.sleep` is called first? We added a latch right before calling `sleep` but that is not deterministic either, however it improves our chances...

**Comment by adamgfraser** [comment]
@joroKr21 I don't think that is correct. The `TestClock` contains specific logic to sure that all fibers in the test have entered a suspended state before advancing the clock. So for example this test passes consistently:

```scala
for {
  ref    <- Ref.make(false)
  _      <- ref.set(true).delay(10.hours).fork
  _      <- adjust(11.hours)
  result <- ref.get
} yield assert(result)(isTrue)
```

**Comment by joroKr21** [comment]
But how do we know if there is isn't some suspension introduced by other means?

We have this test code:
```scala
def process(service: TestService, exit: Exit[String, Int]) = for {
    latch <- CountDownLatch.make(1)
    fiber <- service.processorMetrics.metered(latch.countDown *> ZIO.sleep(duration) *> ZIO.done(exit)).fork
    _ <- latch.await // wait until we call ZIO.sleep before adjusting the clock
    _ <- TestClock.adjust(duration)
    _ <- fiber.join.run
  } yield ()
```

Without the latch it fails, otherwise it works.

**Comment by adamgfraser** [comment]
The `TestClock` monitors all fibers forked within the test and waits for them to complete or enter a suspended state before advancing the test clock. So I'm not sure what is going on in `service.processorMetrics.metered` but if those are normal ZIO effects running on fibers then that should just work.

Notice also that you're latch doesn't actually ensure that `sleep` is called before `adjust` because `latch.countdown` is called before `sleep` and `latchl.await` is called before `adjust`, so there is no actually "happens before" relationship introduced there. That makes me think something else is going on here.

**Comment by joroKr21** [comment]
> Notice also that you're latch doesn't actually ensure that sleep is called before adjust because latch.countdown is called before sleep and latchl.await is called before adjust, so there is no actually "happens before" relationship introduced there. That makes me think something else is going on here.

Yeah exactly, that's what I meant as well when I said it's not really deterministic either but it helps 🤷 

> The TestClock monitors all fibers forked within the test and waits for them to complete or enter a suspended state before advancing the test clock. So I'm not sure what is going on in service.processorMetrics.metered but if those are normal ZIO effects running on fibers then that should just work.

I'm not sure what it means to be normal ZIO effects. To me `Suspended` sounds like it could have many causes, not just waiting for the clock. Anyway `metered` is surrounding the effect with some metrics and logs but I've minimized it to:
```scala
def metered[E, A](
  process: ZIO[AppEnvironment, E, A]
): ZIO[AppEnvironment, E, A] =
  log.info("Staring processor") *> process
```

**Comment by adamgfraser** [comment]
So is it really about this `metered` operator?

What happens if you do this?

```scala
  def process(service: TestService, exit: Exit[String, Int]) = for {
    fiber <- (ZIO.sleep(duration) *> ZIO.done(exit)).fork
    _      <- TestClock.adjust(duration)
    _      <- fiber.join.run
  } yield ()
```

`Suspended` is a specific state that a fiber can have in the runtime. It indicates that the fiber is not doing anything and is waiting for some condition. You're right that it can be many things such as waiting for a `Promise` to be completed or waiting for the clock to be advanced. When all fibers are in a suspended state then nothing else is going to happen in the test until the clock is advanced so it is then safe to advance the clock. There isn't a race condition at that point because there is no one to race against.

**Comment by joroKr21** [comment]
> What happens if you do this?

It doesn't deadlock 🤷 - I couldn't create a standalone reproduction though, sorry

**Comment by adamgfraser** [comment]
Right. That is the expected behavior. The `TestClock` waits for the fibers in the test to enter a suspended state before advancing the clock. I'm not surprised since there are a ton of tests for this. So I think the question is what is going on in this `log.info` call that is throwing things off.

**Comment by joroKr21** [comment]
I tried to create an isolated test that uses only `zio-logging` but I haven't managed yet, for now I have to move on.

**Comment by adamgfraser** [comment]
Okay. Thanks for reporting. Is `zio-logging` what you are using? Happy to take a look with you when you have more time.

**Comment by joroKr21** [comment]
> Okay. Thanks for reporting. Is zio-logging what you are using?

Yes, using `zio-logging`

> When all fibers are in a suspended state then nothing else is going to happen in the test until the clock is advanced so it is then safe to advance the clock. There isn't a race condition at that point because there is no one to race against.

@adamgfraser are you sure it's possible to get a consistent snapshot of the state of all fibers? That sounds like a really expensive piece of global state to keep in sync.

**Comment by adamgfraser** [comment]
@joroKr21 We don't need to look at all the fibers, only the fiber that is executing that test and any fibers that it forks. So in your test for example there are only two fibers. Definitely not something we would want to do in performance sensitive production code but not a problem in tests. This feature has been around for more than a year now and is used pretty extensively so relatively high level of confidence. The main limitations are all of the logic in your test should be executed on ZIO fibers and you shouldn't fork a fiber that never suspends within one of your tests.

**Comment by joroKr21** [comment]
> @joroKr21 We don't need to look at all the fibers, only the fiber that is executing that test and any fibers that it forks.

Ah cool, so it's only synchronised per test 👍 

But, from looking at the code - it seems that the logic to collect fiber state is only applied in `DefaultRunnableSpec` and `MutableRunnableSpec` but not in `RunnableSpec` (which we extend to provide an extended environment). And `TestAspect.fibers` is `private[zio]` so we need to hack around the access restriction by having a `package zio` in our own project 😃

**Comment by reibitto** [comment]
It's been a long time since I've looked at this issue. So my particular issue has to do with `effectBlockingInterrupt` and suspension (or lack of)?

I'm not sure why I made my "minimal" example more complicated than necessary. I think this would have been enough to show it:

```scala
testM("test cancel") {
  val zio = for {
    fiber  <- effectBlockingInterrupt(Thread.sleep(5000)).timeout(3.seconds).fork
    _      <- TestClock.adjust(3.seconds)
    result <- fiber.join
  } yield result

  assertM(zio)(isNone) // will fail
}
```

while this works as you would expect it to:

```scala
testM("test cancel") {
  val zio = for {
    fiber  <- ZIO.sleep(5.seconds).timeout(3.seconds).fork
    _      <- TestClock.adjust(3.seconds)
    result <- fiber.join
  } yield result

  assertM(zio)(isNone)
}
```

**Comment by joroKr21** [comment]
@reibitto it looks like your problem is different because you are using `DefaultRunnableSpec` right?

**Comment by reibitto** [comment]
Yeah, I do use `DefaultRunnableSpec`, but I'm not sure that matters?

**Comment by reibitto** [comment]
Oh, I missed your other comment. I see. These are 2 separate issues maybe.

**Comment by adamgfraser** [comment]
Yes. If the aspect is not being applied at all that would definitely explain the problem you are seeing in your test.

As far as the example where you do `Thread.sleep` I am inclined to say that is just outside the scope of what the test clock is designed to do.

**Comment by joroKr21** [comment]
There are so many issues about hanging `TestClock` so I'm not sure where add this but this test hands when you run it on one processor (emulate with `-XX:ActiveProcessorCount=1`) - `yieldNow` is necessary:

```scala
import zio._
import zio.test._

object Test extends ZIOSpecDefault {

  override def spec = suite("suite")(
    test("test") {
      for {
        _ <- (ZIO.blocking(ZIO.succeed(42)) *> ZIO.yieldNow *> ZIO.sleep(2.seconds)).fork
        _ <- TestClock.adjust(5.seconds)
      } yield assertTrue(true)
    }
  )
}
```
